cmd/gen-ca/main.go

```diff
@@ -0,0 +1,363 @@
+	id, err := hex.DecodeString(idStr)
+	if err != nil {
+		return nil, err
+	}
+
+	privateHandle, err := findObject(ctx, session, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve private key handle: %s", err)
+	}
+	attrs, err := ctx.GetAttributeValue(session, privateHandle, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil)},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve key type: %s", err)
+	}
+	if len(attrs) == 0 {
+		return nil, errors.New("failed to retrieve key attributes")
+	}
+
+	pubHandle, err := findObject(ctx, session, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, attrs[0].Value),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve public key handle: %s", err)
+	}
+	var pub crypto.PublicKey
+	var keyType pkcs11helpers.KeyType
+	switch {
+	// CKK_RSA, 0x00000000
+	case bytes.Compare(attrs[0].Value, []byte{0, 0, 0, 0, 0, 0, 0, 0}) == 0:
+		keyType = pkcs11helpers.RSAKey
+		pub, err = pkcs11helpers.GetRSAPublicKey(ctx, session, pubHandle)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve public key: %s", err)
+		}
+	// CKK_ECDSA, 0x00000003
+	case bytes.Compare(attrs[0].Value, []byte{3, 0, 0, 0, 0, 0, 0, 0}) == 0:
+		keyType = pkcs11helpers.ECDSAKey
+		pub, err = pkcs11helpers.GetECDSAPublicKey(ctx, session, pubHandle)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve public key: %s", err)
+		}
+	default:
+		return nil, errors.New("unsupported key type")
+	}
+
+	return &x509Signer{
+		ctx:          ctx,
+		session:      session,
+		objectHandle: privateHandle,
+		keyType:      keyType,
+		pub:          pub,
+	}, nil
+}
+
+var algToString = map[string]x509.SignatureAlgorithm{
+	"SHA256WithRSA":   x509.SHA256WithRSA,
+	"SHA384WithRSA":   x509.SHA384WithRSA,
+	"SHA512WithRSA":   x509.SHA512WithRSA,
+	"ECDSAWithSHA256": x509.ECDSAWithSHA256,
+	"ECDSAWithSHA384": x509.ECDSAWithSHA384,
+	"ECDSAWithSHA512": x509.ECDSAWithSHA512,
+}
+
+type certProfile struct {
+	SignatureAlgorithm string
+
+	Subject struct {
+		CommonName   string
+		Organization string
+		Country      string
+	}
+
+	NotBefore string
+	NotAfter  string
+
+	OCSPURL   string
+	CRLURL    string
+	IssuerURL string
+
+	PolicyOIDs []string
+}
+
+func parseOID(oidStr string) (asn1.ObjectIdentifier, error) {
+	var oid asn1.ObjectIdentifier
+	for _, a := range strings.Split(oidStr, ".") {
+		i, err := strconv.Atoi(a)
+		if err != nil {
+			return nil, err
+		}
+		oid = append(oid, i)
+	}
+	return oid, nil
+}
+
+const dateLayout = "2006-01-02 15:04:05"
+const serialLength = 16
```

Comment:
I think this `const serialLength` and the `const dateLayout` above could be put inside of the `constructCert` function scope instead of being global.